gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/aggregation/functions/average/AverageProperty.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright © 2014 - 2019 Leipzig University (Database Research Group)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradoop.flink.model.impl.operators.aggregation.functions.average;
+
+import org.gradoop.common.model.api.entities.EPGMElement;
+import org.gradoop.common.model.impl.properties.PropertyValue;
+import org.gradoop.common.model.impl.properties.PropertyValueUtils;
+import org.gradoop.flink.model.api.functions.AggregateDefaultValue;
+import org.gradoop.flink.model.impl.operators.aggregation.functions.BaseAggregateFunction;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Base class for aggregate functions determining the average of a numeric property value.<br>
+ * This aggregate function uses a list of two property values for aggregation internally.
+ * The list will contain two values:
+ * <ol start=0>
+ *   <li>The sum all values considered by the average.</li>
+ *   <li>The number of values added to the sum.</li>
+ * </ol>
+ * A post-processing step is necessary, after the aggregation, to get the final average value.
+ * The final value will be a {@code double} value or {@link PropertyValue#NULL_VALUE null},
```

Comment:
The post-processing step is executed after setting the default value, which in this case returns `NULL_VALUE`.

Reply:
I would prefer doing the post processing only on existing values first and use the default only as default for the whole aggregation, instead of post processing the default value. It is confusing to see the default value and having to apply the post processing myself in my head.

Reply:
The default value should be the default value independent from post-processing. Have a look at the interface `AggregateDefaultValue` ... this can be used for e.g. count to switch from `null` to `0` but not for average. In this case, `getDefaultValue()` should not be overridden. Maybe rename your attribute `defaultValue` to initValue or defaultIncrement ore something else ...